charms/knative-operator/src/charm.py

```diff
@@ -142,6 +186,29 @@ def _on_knative_operator_pebble_ready(self, event):
         self.unit.status = MaintenanceStatus("Configuring Pebble layer")
         self._update_layer(event)
 
+    def _on_otel_collector_relation_changed(self, event):
+        """Event handler for on['otel-collector'].relation_changed."""
+        # Apply all changes only if the otel collector has not been deployed
+        if not self._otel_exporter_ip:
+            self.unit.status = MaintenanceStatus("Applying otel collector")
+            self._template_files_ext = [
+                f"{self._src_dir}/manifests/observability/collector.yaml.j2"
+            ]
+            # Reset the resource handler so it uses the expanded template files list
+            self._resource_handler = None
+            self._apply_all_resources()
```

Comment:
Is there a reason why we want to append this to the rest of the resource handler's templates and (likely re-)run the resource handler if we need the otel deployment?  Does something change about the rest of the resources that needs to be edited here?

What about:
```python
if not self._otel_exporter_ip:
  self.unit.status = MaintenanceStatus(...)
  self._otel_resource_handler.apply()
  # or... self._apply_otel_resources(), which can have the error handling for nice messages
```

I understand the draw of reusing the code from _apply_all_resources, but it couples two things that don't need to be coupled afaict.  With them coupled, unit testing becomes harder because of the interactions between these.

Reply:
I have added an observability specific resource handler, let me know what you think.

Reply:
yeah this looks way better.  love it